// @calimero-network/calimero-client@1.14.2/lib/api/contractApi.js
export var ActionScope;
(function (ActionScope) {
    ActionScope["ExternalFunctionCall"] = "ExternalFunctionCall";
    ActionScope["Transfer"] = "Transfer";
    ActionScope["SetNumApprovals"] = "SetNumApprovals";
    ActionScope["SetActiveProposalsLimit"] = "SetActiveProposalsLimit";
    ActionScope["SetContextValue"] = "SetContextValue";
    ActionScope["DeleteProposal"] = "DeleteProposal";
})(ActionScope || (ActionScope = {}));
export function createExternalFunctionCall(receiver_id, method_name, args, deposit) {
    return {
        scope: ActionScope.ExternalFunctionCall,
        params: {
            receiver_id,
            method_name,
            args,
            deposit,
        },
    };
}
export function createTransfer(receiver_id, amount) {
    return {
        scope: ActionScope.Transfer,
        params: {
            receiver_id,
            amount,
        },
    };
}
export function createSetNumApprovals(num_approvals) {
    return {
        scope: ActionScope.SetNumApprovals,
        params: {
            num_approvals,
        },
    };
}
export function createSetActiveProposalsLimit(active_proposals_limit) {
    return {
        scope: ActionScope.SetActiveProposalsLimit,
        params: {
            active_proposals_limit,
        },
    };
}
export function createSetContextValue(key, value) {
    return {
        scope: ActionScope.SetContextValue,
        params: {
            key,
            value,
        },
    };
}
export function createDeleteProposal(proposal_id) {
    return {
        scope: ActionScope.DeleteProposal,
        params: {
            proposal_id,
        },
    };
}
