// @calimero-network/calimero-client@1.14.2/lib/api/index.js
import axios from 'axios';
import { AxiosHttpClient } from './httpClient';
import { NodeApiDataSource } from './dataSource/NodeApiDataSource';
import { AuthApiDataSource } from './dataSource/AuthApiDataSource';
import { ContractApiDataSource } from './dataSource/ContractApiDataSource';
import { JsonRpcClient } from '../rpc/jsonrpc';
import { AdminApiDataSource } from './dataSource/AdminApiDataSource';
import { BlobApiDataSource } from './dataSource/BlobApiDataSource';
class ApiClient {
    nodeApi;
    authApi;
    contractApi;
    adminApi;
    blobApi;
    jsonRpcClient;
    constructor(httpClient) {
        this.nodeApi = new NodeApiDataSource(httpClient);
        this.authApi = new AuthApiDataSource(httpClient);
        this.contractApi = new ContractApiDataSource(httpClient);
        this.adminApi = new AdminApiDataSource(httpClient);
        this.blobApi = new BlobApiDataSource(httpClient);
        this.jsonRpcClient = new JsonRpcClient(httpClient);
    }
    node() {
        return this.nodeApi;
    }
    auth() {
        return this.authApi;
    }
    contract() {
        return this.contractApi;
    }
    admin() {
        return this.adminApi;
    }
    blob() {
        return this.blobApi;
    }
    rpc() {
        return this.jsonRpcClient;
    }
}
const apiClient = new ApiClient(new AxiosHttpClient(axios));
const authClient = new AuthApiDataSource(new AxiosHttpClient(axios));
const contractClient = new ContractApiDataSource(new AxiosHttpClient(axios));
const adminClient = new AdminApiDataSource(new AxiosHttpClient(axios));
const blobClient = new BlobApiDataSource(new AxiosHttpClient(axios));
const rpcClient = new JsonRpcClient(new AxiosHttpClient(axios));
export { apiClient, authClient, contractClient, adminClient, rpcClient, blobClient, };
