// @calimero-network/calimero-client@1.14.2/lib/api/dataSource/NodeApiDataSource.js
import { getAppEndpointKey } from '../../storage';
export class NodeApiDataSource {
    client;
    constructor(client) {
        this.client = client;
    }
    get baseUrl() {
        return getAppEndpointKey();
    }
    async health(request) {
        return await this.client.get(`${request.url}/admin-api/health`);
    }
    async getContext(contextId) {
        try {
            return this.client.get(`${this.baseUrl}/admin-api/contexts/${contextId}`);
        }
        catch (error) {
            console.error('Error fetching context:', error);
            return {
                data: null,
                error: { code: 500, message: 'Failed to fetch context data.' },
            };
        }
    }
    async getContexts() {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin-api/contexts`);
            return response;
        }
        catch (error) {
            console.error('Error fetching contexts:', error);
            return { error: { code: 500, message: 'Failed to fetch context data.' } };
        }
    }
    async createContext(applicationId, jsonParams, protocol) {
        const initializationParams = jsonParams === '' ? [] : Array.from(new TextEncoder().encode(jsonParams));
        try {
            const response = await this.client.post(`${this.baseUrl}/admin-api/contexts`, {
                applicationId,
                initializationParams,
                protocol,
            });
            return response;
        }
        catch (error) {
            console.error('Error starting contexts:', error);
            return { error: { code: 500, message: 'Failed to start context.' } };
        }
    }
    async deleteContext(contextId) {
        try {
            const response = await this.client.delete(`${this.baseUrl}/admin-api/contexts/${contextId}`);
            return response;
        }
        catch (error) {
            console.error('Error deleting context:', error);
            return { error: { code: 500, message: 'Failed to delete context.' } };
        }
    }
    async fetchContextIdentities(contextId) {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin-api/contexts/${contextId}/identities-owned`);
            return response;
        }
        catch (error) {
            console.error('Error deleting context:', error);
            return { error: { code: 500, message: 'Failed to delete context.' } };
        }
    }
    async createNewIdentity() {
        try {
            const response = await this.client.post(`${this.baseUrl}/admin-api/identity/context`);
            return response;
        }
        catch (error) {
            console.error('Error creating new identity:', error);
            return {
                error: { code: 500, message: 'Failed to create new identity.' },
            };
        }
    }
    async contextInvite(contextId, inviterPublicKey, inviteePublicKey) {
        try {
            const response = await this.client.post(`${this.baseUrl}/admin-api/contexts/invite`, {
                contextId: contextId,
                inviterId: inviterPublicKey,
                inviteeId: inviteePublicKey,
            });
            return response;
        }
        catch (error) {
            console.error('Error inviting to context:', error);
            return { error: { code: 500, message: 'Failed to invite to context.' } };
        }
    }
    async joinContext(invitationPayload) {
        try {
            const response = await this.client.post(`${this.baseUrl}/admin-api/contexts/join`, {
                invitationPayload,
            });
            return response;
        }
        catch (error) {
            console.error('Error joining context:', error);
            return { error: { code: 500, message: 'Failed to join context.' } };
        }
    }
    // Application Management
    async getInstalledApplications() {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin-api/applications`);
            return response;
        }
        catch (error) {
            console.error('Error fetching installed applications:', error);
            return {
                error: {
                    code: 500,
                    message: 'Failed to fetch installed applications.',
                },
            };
        }
    }
    async getInstalledApplicationDetails(appId) {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin-api/applications/${appId}`);
            return response;
        }
        catch (error) {
            console.error('Error fetching application details:', error);
            return {
                error: { code: 500, message: 'Failed to fetch application details.' },
            };
        }
    }
    async installApplication(url, metadata, hash) {
        try {
            const requestBody = {
                url,
                metadata: metadata ? Array.from(metadata) : [],
                ...(hash ? { hash } : {}),
            };
            const response = await this.client.post(`${this.baseUrl}/admin-api/install-application`, requestBody);
            return response;
        }
        catch (error) {
            console.error('Error installing application:', error);
            return {
                error: { code: 500, message: 'Failed to install application.' },
            };
        }
    }
    async uninstallApplication(applicationId) {
        try {
            const response = await this.client.delete(`${this.baseUrl}/admin-api/applications/${applicationId}`);
            return response;
        }
        catch (error) {
            console.error('Error uninstalling application:', error);
            return {
                error: { code: 500, message: 'Failed to uninstall application.' },
            };
        }
    }
    // Context Management Extended
    async getContextClientKeys(contextId) {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin-api/contexts/${contextId}/client-keys`);
            return response;
        }
        catch (error) {
            console.error('Error fetching context client keys:', error);
            return {
                error: { code: 500, message: 'Failed to fetch context client keys.' },
            };
        }
    }
    async getContextUsers(contextId) {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin-api/contexts/${contextId}/identities`);
            return response;
        }
        catch (error) {
            console.error('Error fetching context users:', error);
            return {
                error: { code: 500, message: 'Failed to fetch context users.' },
            };
        }
    }
    async getContextStorageUsage(contextId) {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin-api/contexts/${contextId}/storage`);
            return response;
        }
        catch (error) {
            console.error('Error fetching context storage usage:', error);
            return {
                error: { code: 500, message: 'Failed to fetch context storage usage.' },
            };
        }
    }
    async grantCapabilities(contextId, request) {
        try {
            const response = await this.client.post(`${this.baseUrl}/admin-api/contexts/${contextId}/capabilities/grant`, request);
            return response;
        }
        catch (error) {
            console.error('Error granting capabilities:', error);
            return { error: { code: 500, message: 'Failed to grant capabilities.' } };
        }
    }
    async revokeCapabilities(contextId, request) {
        try {
            const response = await this.client.post(`${this.baseUrl}/admin-api/contexts/${contextId}/capabilities/revoke`, request);
            return response;
        }
        catch (error) {
            console.error('Error revoking capabilities:', error);
            return {
                error: { code: 500, message: 'Failed to revoke capabilities.' },
            };
        }
    }
    async checkAuth() {
        const response = await this.client.get(`${this.baseUrl}/admin-api/is-authed`);
        return response;
    }
}
