// @calimero-network/calimero-client@1.14.2/lib/experimental/CalimeroProvider.js
import React, { createContext, useContext, useState, useEffect, useCallback, useMemo, } from 'react';
import { apiClient } from '../api';
import { clientLogout, getAccessToken, getAppEndpointKey, setAccessToken, setAppEndpointKey, setRefreshToken, } from '../storage/storage';
import CalimeroLoginModal from './CalimeroLoginModal';
import Toast from './Toast';
import { CalimeroApplication } from './app';
import { AppMode } from './types';
import { GlobalStyle } from '../styles/global';
const CalimeroContext = createContext({
    app: null,
    isAuthenticated: false,
    login: () => { },
    logout: () => { },
    appUrl: null,
    isOnline: true,
});
export const useCalimero = () => useContext(CalimeroContext);
const getPermissionsForMode = (mode) => {
    switch (mode) {
        case AppMode.MultiContext:
            return ['context:execute', 'application'];
        default:
            throw new Error(`Unsupported application mode: ${mode}`);
    }
};
export const CalimeroProvider = ({ children, clientApplicationId, mode, applicationPath, }) => {
    const [isAuthenticated, setIsAuthenticated] = useState(false);
    const [isLoginOpen, setIsLoginOpen] = useState(false);
    const [isLoading, setIsLoading] = useState(true);
    const [toast, setToast] = useState(null);
    const [isOnline, setIsOnline] = useState(true);
    const [appUrl, setAppUrl] = useState(getAppEndpointKey());
    const performLogin = useCallback((url) => {
        const permissions = getPermissionsForMode(mode);
        apiClient.auth().login({
            url,
            callbackUrl: window.location.href,
            applicationId: clientApplicationId,
            permissions,
            applicationPath,
        });
    }, [clientApplicationId, mode, applicationPath]);
    const logout = useCallback(() => {
        clientLogout();
        setIsAuthenticated(false);
        setIsOnline(true);
        setAppUrl(null);
    }, []);
    useEffect(() => {
        const fragment = window.location.hash.substring(1); // Remove the leading #
        const fragmentParams = new URLSearchParams(fragment);
        const encodedAccessToken = fragmentParams.get('access_token');
        const encodedRefreshToken = fragmentParams.get('refresh_token');
        if (encodedAccessToken && encodedRefreshToken) {
            window.history.replaceState({}, document.title, window.location.pathname);
            const accessToken = decodeURIComponent(encodedAccessToken);
            const refreshToken = decodeURIComponent(encodedRefreshToken);
            setAccessToken(accessToken);
            setRefreshToken(refreshToken);
            const newAppUrl = getAppEndpointKey();
            setAppUrl(newAppUrl);
            if (!newAppUrl)
                return;
            const verify = async () => {
                const response = await apiClient.node().checkAuth();
                if (!response.error) {
                    setIsAuthenticated(true);
                }
            };
            verify();
        }
    }, []);
    useEffect(() => {
        const checkSession = async () => {
            const savedUrl = getAppEndpointKey();
            const savedToken = getAccessToken();
            if (savedUrl && savedToken) {
                try {
                    const response = await apiClient.node().checkAuth();
                    if (!response.error) {
                        setIsAuthenticated(true);
                        setIsOnline(true);
                    }
                    else if (response.error.code === 401) {
                        performLogin(savedUrl);
                    }
                }
                catch (error) {
                    logout();
                }
            }
            setIsLoading(false);
        };
        checkSession();
    }, [performLogin, logout]);
    useEffect(() => {
        const intervalId = setInterval(async () => {
            if (!isAuthenticated)
                return;
            const savedUrl = getAppEndpointKey();
            if (savedUrl) {
                try {
                    const response = await apiClient.node().checkAuth();
                    if (response.error && response.error.code === 401) {
                        logout();
                        setToast({
                            message: 'Session expired. Please connect again.',
                            type: 'error',
                        });
                        return;
                    }
                    if (response.error && isOnline) {
                        setToast({
                            message: 'Connection lost. Trying to reconnect...',
                            type: 'error',
                        });
                        setIsOnline(false);
                    }
                    else if (!response.error && !isOnline) {
                        setToast({ message: 'Connection restored.', type: 'success' });
                        setIsOnline(true);
                        setTimeout(() => setToast(null), 5000);
                    }
                }
                catch (error) {
                    if (isOnline) {
                        setToast({
                            message: 'Connection lost. Trying to reconnect...',
                            type: 'error',
                        });
                        setIsOnline(false);
                    }
                }
            }
        }, 5000);
        return () => clearInterval(intervalId);
    }, [isAuthenticated, isOnline, logout]);
    const handleConnect = (url) => {
        setAppEndpointKey(url);
        setAppUrl(url);
        performLogin(url);
    };
    const login = () => setIsLoginOpen(true);
    const app = useMemo(() => isAuthenticated
        ? new CalimeroApplication(apiClient, clientApplicationId)
        : null, [isAuthenticated, clientApplicationId]);
    useEffect(() => {
        // Closes WebSocket connection on logout or when the component unmounts
        return () => {
            app?.close();
        };
    }, [app]);
    return (React.createElement(CalimeroContext.Provider, { value: { app, isAuthenticated, login, logout, appUrl, isOnline } }, isLoading ? (React.createElement("div", null, "Loading...")) : (React.createElement(React.Fragment, null,
        React.createElement(GlobalStyle, null),
        children,
        isLoginOpen && (React.createElement(CalimeroLoginModal, { onConnect: handleConnect, onClose: () => setIsLoginOpen(false) })),
        toast && (React.createElement(Toast, { message: toast.message, type: toast.type, onClose: () => setToast(null) }))))));
};
