// .ignored/@calimero-network/calimero-client/lib/login/ClientLogin.js
import React, { useCallback, useEffect, useState } from 'react';
import { clearAccessToken, clearAppEndpoint, clearApplicationId, clearContextId, clearExecutorPublicKey, getAppEndpointKey, getApplicationId, setContextId, setExecutorPublicKey, } from '../storage/storage';
import { ErrorMessage, LoginButton, LoginHeader, LoginHeaderSpan, Button, ModalOverlay, ModalContent, } from './Components';
import { apiClient } from '../api';
import Spinner from '../components/loader/Spinner';
import { SetupSpinnerContainer } from '../setup/Components';
import { SelectContext } from './noAuth/SelectContext';
import { SelectContextIdentity } from './noAuth/SelectContextIdentity';
const initialState = {
    nodeServerUrl: '',
    applicationId: '',
    selectedContextId: '',
    contexts: [],
    contextIdentities: [],
    errorMessage: '',
    isLoading: false,
};
export const ClientLogin = ({ permissions, authMode, setIsAuthenticated, clientApplicationId, clientApplicationPath, fetchContextApplication, onReset, }) => {
    const [state, setState] = useState({
        ...initialState,
        nodeServerUrl: getAppEndpointKey() ?? '',
        applicationId: getApplicationId() ?? '',
    });
    const { nodeServerUrl, selectedContextId, contexts, contextIdentities, errorMessage, isLoading, } = state;
    const updateState = (newState) => {
        setState((prev) => ({ ...prev, ...newState }));
    };
    const resetSetup = () => {
        clearAppEndpoint();
        clearAccessToken();
        clearContextId();
        clearExecutorPublicKey();
        clearApplicationId();
        onReset();
    };
    const fetchAvailableContexts = useCallback(async () => {
        if (errorMessage)
            return;
        try {
            const response = await apiClient
                .node()
                .getContexts();
            if (response.error) {
                updateState({ errorMessage: response.error.message, isLoading: false });
                return;
            }
            updateState({ contexts: response.data?.contexts, isLoading: false });
        }
        catch (error) {
            updateState({
                errorMessage: 'Failed to fetch contexts',
                isLoading: false,
            });
        }
    }, [errorMessage]);
    const fetchContextIdentities = useCallback(async () => {
        if (!selectedContextId)
            return;
        try {
            const response = await apiClient.node().fetchContextIdentities(selectedContextId);
            if (response.error) {
                updateState({ errorMessage: response.error.message });
                return;
            }
            updateState({ contextIdentities: response.data.identities });
        }
        catch (error) {
            updateState({ errorMessage: 'Failed to fetch context identities' });
        }
    }, [selectedContextId]);
    const handleIdentitySelection = (contextId, identity) => {
        setContextId(contextId);
        setExecutorPublicKey(identity);
        fetchContextApplication();
    };
    const login = useCallback(async () => {
        apiClient.auth().login({
            url: nodeServerUrl,
            callbackUrl: window.location.href,
            permissions: permissions,
            applicationId: clientApplicationId,
            applicationPath: clientApplicationPath,
        });
    }, [nodeServerUrl, permissions, clientApplicationId, clientApplicationPath]);
    useEffect(() => {
        if (authMode === false) {
            updateState({ isLoading: true });
            fetchAvailableContexts();
        }
    }, [nodeServerUrl, fetchAvailableContexts, authMode]);
    useEffect(() => {
        if (selectedContextId) {
            fetchContextIdentities();
        }
    }, [selectedContextId, fetchContextIdentities]);
    const renderLoginContent = () => {
        if (isLoading) {
            return (React.createElement(SetupSpinnerContainer, null,
                React.createElement(Spinner, null)));
        }
        if (authMode === true) {
            return (React.createElement(React.Fragment, null,
                React.createElement(LoginButton, { onClick: login }, "Login"),
                React.createElement(Button, { onClick: resetSetup, style: { marginTop: '1rem' } }, "Back to Setup")));
        }
        return (React.createElement(React.Fragment, null,
            React.createElement(LoginHeader, null,
                React.createElement(LoginHeaderSpan, null, "Select Context ID and Identity")),
            !selectedContextId ? (React.createElement(SelectContext, { contextList: contexts, setSelectedContextId: (id) => updateState({ selectedContextId: id }) })) : (React.createElement(SelectContextIdentity, { selectedContextId: selectedContextId, contextIdentities: contextIdentities, onSelectIdentity: handleIdentitySelection, backStep: () => updateState({ selectedContextId: '' }) })),
            React.createElement(Button, { onClick: resetSetup, style: { marginTop: '1rem' } }, "Back to Setup")));
    };
    return (React.createElement(ModalOverlay, null,
        React.createElement(ModalContent, null,
            errorMessage && (React.createElement(React.Fragment, null,
                React.createElement(ErrorMessage, null, errorMessage),
                React.createElement(Button, { onClick: resetSetup, style: { marginTop: '1rem' } }, "Back to Setup"))),
            renderLoginContent())));
};
