// .ignored/@calimero-network/calimero-client/lib/api/dataSource/AdminApiDataSource.js
import { getAppEndpointKey } from '../../storage';
export class AdminApiDataSource {
    client;
    constructor(client) {
        this.client = client;
    }
    get baseUrl() {
        return getAppEndpointKey();
    }
    async getRootKeys() {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin/keys`);
            return response;
        }
        catch (error) {
            console.error('Error fetching root keys:', error);
            return { error: { code: 500, message: 'Failed to fetch root keys.' } };
        }
    }
    async getClientKeys() {
        try {
            const response = await this.client.get(`${this.baseUrl}/admin/keys/clients`);
            return response;
        }
        catch (error) {
            console.error('Error fetching client keys:', error);
            return { error: { code: 500, message: 'Failed to fetch client keys.' } };
        }
    }
    async addRootKey(rootKeyRequest) {
        try {
            const response = await this.client.post(`${this.baseUrl}/admin/keys`, rootKeyRequest);
            return response;
        }
        catch (error) {
            console.error('Error adding root key:', error);
            return { error: { code: 500, message: 'Failed to add root key.' } };
        }
    }
    async revokeRootKey(keyId) {
        try {
            const response = await this.client.delete(`${this.baseUrl}/admin/keys/${keyId}`);
            return response;
        }
        catch (error) {
            console.error('Error deleting root key:', error);
            return { error: { code: 500, message: 'Failed to delete root key.' } };
        }
    }
    async revokeClientKey(rootKeyId, clientId) {
        try {
            const response = await this.client.delete(`${this.baseUrl}/admin/keys/${rootKeyId}/clients/${clientId}`);
            return response;
        }
        catch (error) {
            console.error('Error revoking client key:', error);
            return { error: { code: 500, message: 'Failed to revoke client key.' } };
        }
    }
    async setKeyPermissions(keyId, request) {
        try {
            const response = await this.client.put(`${this.baseUrl}/admin/keys/${keyId}/permissions`, request);
            return response;
        }
        catch (error) {
            console.error('Error updating key permissions:', error);
            return {
                error: { code: 500, message: 'Failed to update key permissions.' },
            };
        }
    }
}
