// .ignored/@calimero-network/calimero-client/lib/api/dataSource/AuthApiDataSource.js
import { getAppEndpointKey, APP_URL, APPLICATION_ID, } from '../../storage/storage';
export class AuthApiDataSource {
    client;
    constructor(client) {
        this.client = client;
    }
    get baseUrl() {
        return getAppEndpointKey();
    }
    async login(request) {
        try {
            window.location.href = `${request.url}/auth/login?callback-url=${encodeURIComponent(request.callbackUrl)}&permissions=${encodeURIComponent(request.permissions.join(','))}&${APPLICATION_ID}=${encodeURIComponent(request.applicationId)}&application-path=${encodeURIComponent(request.applicationPath)}&${APP_URL}=${encodeURIComponent(this.baseUrl)}`;
            return { data: null }; // The response doesn't matter as we're redirecting
        }
        catch (error) {
            console.error('Error during login redirect:', error);
            return {
                error: {
                    code: 500,
                    message: 'Login redirect failed. Please try again.',
                },
            };
        }
    }
    async refreshToken(request) {
        try {
            const response = await this.client.post(`${this.baseUrl}/auth/refresh`, request);
            return response;
        }
        catch (error) {
            console.error('Error refreshing token:', error);
            return {
                error: {
                    code: 500,
                    message: 'Token refresh failed. Please try again.',
                },
            };
        }
    }
    async getProviders() {
        try {
            const response = await this.client.get(`${this.baseUrl}/auth/providers`);
            return response;
        }
        catch (error) {
            console.error('Error getting providers:', error);
            return { error: { code: 500, message: 'Failed to get providers.' } };
        }
    }
    async requestToken(requestBody) {
        try {
            const response = await this.client.post(`${this.baseUrl}/auth/token`, requestBody);
            return response;
        }
        catch (error) {
            console.error('Error requesting token:', error);
            return {
                error: {
                    code: 500,
                    message: 'Token request failed. Please try again.',
                },
            };
        }
    }
    async getChallenge() {
        try {
            const response = await this.client.get(`${this.baseUrl}/auth/challenge`);
            return response;
        }
        catch (error) {
            console.error('Error getting challenge:', error);
            return { error: { code: 500, message: 'Failed to get challenge.' } };
        }
    }
    async generateClientKey(request) {
        try {
            const response = await this.client.post(`${this.baseUrl}/admin/client-key`, request);
            return response;
        }
        catch (error) {
            console.error('Error generating client key:', error);
            return {
                error: { code: 500, message: 'Failed to generate client key.' },
            };
        }
    }
}
