// .ignored/@calimero-network/calimero-client/lib/api/dataSource/ContractApiDataSource.js
import { getAppEndpointKey, getContextId } from '../../storage/storage';
export class ContractApiDataSource {
    client;
    constructor(client) {
        this.client = client;
    }
    get baseUrl() {
        return getAppEndpointKey();
    }
    get contextId() {
        const id = getContextId();
        if (!id) {
            throw new Error('Context ID not available. Make sure you are properly authenticated.');
        }
        return id;
    }
    async getProposals(request) {
        try {
            return await this.client.post(`${this.baseUrl}/admin-api/contexts/${this.contextId}/proposals`, request);
        }
        catch (error) {
            return {
                data: null,
                error: error,
            };
        }
    }
    async getProposalApprovers(proposalId) {
        try {
            return await this.client.get(`${this.baseUrl}/admin-api/contexts/${this.contextId}/proposals/${proposalId}/approvals/users`);
        }
        catch (error) {
            return {
                data: null,
                error: error,
            };
        }
    }
    async getProposalApprovalCount(proposalId) {
        try {
            return await this.client.get(`${this.baseUrl}/admin-api/contexts/${this.contextId}/proposals/${proposalId}/approvals/count`);
        }
        catch (error) {
            return {
                data: null,
                error: error,
            };
        }
    }
    async getNumOfProposals() {
        try {
            return await this.client.get(`${this.baseUrl}/admin-api/contexts/${this.contextId}/proposals/count`);
        }
        catch (error) {
            return {
                data: null,
                error: error,
            };
        }
    }
    async getProposalDetails(proposalId) {
        try {
            return await this.client.get(`${this.baseUrl}/admin-api/contexts/${this.contextId}/proposals/${proposalId}`);
        }
        catch (error) {
            return {
                data: null,
                error: error,
            };
        }
    }
    async getContextValue(key) {
        try {
            return await this.client.post(`${this.baseUrl}/admin-api/contexts/${this.contextId}`, {
                key,
            });
        }
        catch (error) {
            return {
                data: null,
                error: error,
            };
        }
    }
    async getContextStorageEntries(offset, limit) {
        try {
            return await this.client.post(`${this.baseUrl}/admin-api/contexts/${this.contextId}/proposals/context-storage-entries`, {
                offset,
                limit,
            });
        }
        catch (error) {
            return {
                data: null,
                error: error,
            };
        }
    }
}
